ui/components/src/TimeSeriesTooltip/nearby-series.ts

```diff
@@ -0,0 +1,261 @@
+// Copyright 2023 The Perses Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import { ECharts as EChartsInstance } from 'echarts/core';
+import { formatValue, UnitOptions, EChartsDataFormat, OPTIMIZED_MODE_SERIES_LIMIT } from '../model';
+import { CursorData } from './tooltip-model';
+
+// increase multipliers to show more series in tooltip
+export const INCREASE_NEARBY_SERIES_MULTIPLIER = 5.5; // adjusts how many series show in tooltip (higher == more series shown)
+export const DYNAMIC_NEARBY_SERIES_MULTIPLIER = 30; // used for adjustment after series number divisor
+export const SHOW_FEWER_SERIES_LIMIT = 5;
+
+export interface NearbySeriesInfo {
+  seriesIdx: number | null;
+  datumIdx: number | null;
+  seriesName: string;
+  date: number;
+  markerColor: string;
+  x: number;
+  y: number;
+  formattedY: string;
+  isClosestToCursor: boolean;
+}
+
+export type NearbySeriesArray = NearbySeriesInfo[];
+
+/**
+ * Returns formatted series data for the points that are close to the user's cursor
+ * Adjust yBuffer to increase or decrease number of series shown
+ */
+export function checkforNearbySeries(
+  data: EChartsDataFormat,
+  pointInGrid: number[],
+  yBuffer: number,
+  chart?: EChartsInstance,
+  unit?: UnitOptions
+): NearbySeriesArray {
+  const currentNearbySeriesData: NearbySeriesArray = [];
+  const cursorX: number | null = pointInGrid[0] ?? null;
+  const cursorY: number | null = pointInGrid[1] ?? null;
+
+  if (cursorX === null || cursorY === null) {
+    return currentNearbySeriesData;
+  }
+
+  const nearbySeriesIndexes: number[] = [];
+  const emphasizedSeriesIndexes: number[] = [];
+  const nonEmphasizedSeriesIndexes: number[] = [];
+  const totalSeries = data.timeSeries.length;
+  if (Array.isArray(data.xAxis) && Array.isArray(data.timeSeries)) {
```

Comment:
Not for you to fix in this PR because I know I'm only seeing this because git did a rename, but this deeply nested set of loops and conditionals is kinda hard to read. It sounds like you might touch this again for some follow-ups. If so, it'd be awesome if this could potentially get broken out into some smaller methods that are a little easier to read & maintain.

Reply:
Yeah for sure, planning to add more early returns for the nested if's, separating out utils, and adding more unit tests!

I'd love to write an e2e for tooltip behavior too instead of just the RTL TooltipContent tests, I briefly tried before but I need to level up my Playwright knowledge before trying again